gnuradio-runtime/lib/hier_block2_detail.cc

```diff
@@ -927,5 +937,77 @@ std::string hier_block2_detail::log_level()
     return tmp[0]->log_level();
 }
 
+void hier_block2_detail::set_parent(hier_block2* parent)
+{
+    auto old_parent = d_parent.lock();
+    if (!old_parent) {
+        d_parent = parent->to_hier_block2();
+        d_parent_refcnt++;
+        return;
+    }
+    if (old_parent.get() == parent) {
+        d_parent_refcnt++;
+        return;
+    }
+    std::stringstream msg;
+    msg << "Hierarchical blocks do not yet support moving from one hierarchy "
+           "to another. Block \""
+        << d_owner->name() << "\" moved from \"" << old_parent->name() << "\" to \""
+        << parent->name() << "\"";
+    throw std::runtime_error(msg.str());
```

Comment:
Throwing here looks like it could break existing applications.

Reply:
Throwing may occur only if some block is connected inside one hier_block2 and then connected inside another hier_block2 without being fully disconnected from first hier_block2. I don't think this to be correct usage of hier_block2.
Exception test:
```
#include <iostream>

#include <gnuradio/blocks/null_sink.h>
#include <gnuradio/blocks/null_source.h>
#include <gnuradio/hier_block2.h>
#include <gnuradio/blocks/copy.h>
#include <gnuradio/top_block.h>


class test_block : public gr::hier_block2
{

public:
typedef std::shared_ptr<test_block> sptr;
    static test_block::sptr make()
    {
        return gnuradio::get_initial_sptr(new test_block());
    }
    test_block();
    virtual ~test_block();

    private:
    gr::blocks::copy::sptr d_copy;
};

test_block::test_block(): gr::hier_block2 ("test_block",
                          gr::io_signature::make (1, 1, sizeof(gr_complex)),
                          gr::io_signature::make (1, 1, sizeof(gr_complex)))

{
    d_copy = gr::blocks::copy::make(sizeof(gr_complex));
    connect(self() , 0 , d_copy , 0);
    connect(d_copy , 0, self() , 0);
}

test_block::~test_block()
{
}

class test_block2 : public gr::hier_block2
{

public:
typedef std::shared_ptr<test_block2> sptr;
    static test_block2::sptr make()
    {
        return gnuradio::get_initial_sptr(new test_block2());
    }
    test_block2();
    virtual ~test_block2();

    test_block::sptr d_test;
};

test_block2::test_block2(): gr::hier_block2 ("test_block",
                          gr::io_signature::make (1, 1, sizeof(gr_complex)),
                          gr::io_signature::make (1, 1, sizeof(gr_complex)))

{
    d_test = test_block::make();
    connect(self() , 0 , d_test , 0);
    connect(d_test , 0, self() , 0);
}

test_block2::~test_block2()
{
}

int main()
{
    auto tb = gr::make_top_block("buggy");
    auto test0 = test_block2::make();
    auto test1 = test_block2::make();
    auto src0 = gr::blocks::null_source::make(sizeof(gr_complex));
    auto snk0 = gr::blocks::null_sink::make(sizeof(gr_complex));
    auto snk1 = gr::blocks::null_sink::make(sizeof(gr_complex));
    auto bad_copy = gr::blocks::copy::make(sizeof(gr_complex));
    tb->connect(src0, 0, test0, 0);
    tb->connect(src0, 0, test1, 0);
    tb->connect(test0, 0, snk0, 0);
    tb->connect(test1, 0, snk1, 0);
    test0->disconnect(test0->d_test, 0, test0->self(), 0);
    test1->disconnect(test1->d_test, 0, test1->self(), 0);
    test0->connect(test1->d_test, 0, test0->self(), 0);
    test1->connect(test0->d_test, 0, test1->self(), 0);
    tb->start();
    return 0;
}
```

Reply:
Got it. I think you're correct that an instance of a hierarchical block should only be present inside one parent hierarchical block at a time. I was confused by the error message though:

> Hierarchical blocks do not yet support moving from one hierarchy to another.

This makes it sound like you're not allowed to remove a hierarchical block from one parent and then add it to another. It also makes it sound like this feature will be allowed in the future. How about something like:

> A hierarchical block cannot have multiple parents. Block X must be removed from parent Y before being added to parent Z.

Reply:
I agree. The message is left here from the first version, that had no reference counting implemented. I'll update the message and add 2 tests. Lock on unconnected hier_block2 and lock after disconnection.